proposals/format.md

```diff
@@ -0,0 +1,320 @@
+The CoreFX is introducing a new managed type `Variant`. This type is meant to be used in APIs which expect hetrogeneous
+values but don't want the overhead brought on by using `object` as the parmeter. The `Variant` type provides universal 
+storage but avoids the boxing allocation for the most commonly used types. Using this type in APIs like `string.Format`
+can eliminate the boxing overhead in the majority of cases.
+
+This type itself is not necessarily special to the language. It is being introduced in this document separately though
+as it becomes an implementation detail of other parts of the proposal. 
+
+### Efficient interpolated strings
+Interpolated strings are a popular yet innefecient feature in C#. The most common syntax, using an interpolated `string`
+as a `string`, translates into a `string.Format(string, params object[])` call. That will inccur boxing allocations for 
+all value types, intermediate `string` allocations as the implementation largely uses `object.ToString` for formatting
+as well as array allocations once the number of arguments exceeds the amount of parameters on the "fast" overloads of 
+`string.Format`. 
+
+The language will change it's interpolation lowering to consider alternate overloads of `string.Format`. It will
+consider all forms of `string.Format(string, params)` and pick the "best" overload which satisfies the argument types.
+The "best" `params` overload will be determined by the rules discussed above. This means interpolated `string` can now
+bind to very efficient overloads like `string.Format(string format, params ReadOnlySpan<Variant> args)`. In many cases
+this will remove all intermediate allocations.
+
+### Customizable interpolated strings
+Developers are able to customize the behavior of interpolated strings with `FormattableString`. This contains the data
+which goes into an interpolated string: the format `string` and the arguments as an array. This though still has the 
+boxing and argument array allocation as well as the allocation for `FormattableString` (it's an `abstract class`). Hence
+it's of little use to applications which are allocation heavy in `string` formatting.
+
+To make interopolated string formatting efficient the language will recognize a new type: 
+`System.ValueFormattableString`. All interpolated strings will have a target type conversion to this type. This will 
+be implemented by translating the interpolated string into the call `ValueFormattableString.Create` exactly as is done
+for `FormattableString.Create` today. The language will support all `params` options described in this document when
+looking for the most suitable `ValueFormattableString.Create` method. 
+
+``` csharp
+readonly struct ValueFormattableString {
+    public static ValueFormattableString Create(Variant v) { ... } 
+    public static ValueFormattableString Create(string s) { ... } 
+    public static ValueFormattableString Create(string s, params ReadOnlySpan<Variant> collection) { ... } 
```

Comment:
would you need a ReadOnlySpan of hte string portions of hte interpolated string, along with the ReadOnlySPan of interpolation values?  Or am i misunderstanding how this would be used?  For example, if you have:

`$"foo{bar}baz"`

how is the ValueFormattableString created?  I would prefer a form that gave me `"foo"` and `"bar"` in a ReadOnlySpan, and something that gave me the Variant for `bar` in a ReadOnlySpan.